### PR TITLE
[#105]  Align Kafka Extension page

### PR DESCRIPTION
**Pull request by smcvb**
This pull request updates the entire `kafka.md` file to accommodate for all the recent changes which have been made in the [Kafka Extension](https://github.com/AxonFramework/extension-kafka).
As such it resolves issue #105

### CODE REVIEW

File: extensions/kafka.md [hunk 1]
```diff
@@ -1,91 +1,437 @@
...
-Kafka is an alternative approach to distributing events, besides Axon Server which is default.
+{% hint style="info" %}
+
+The Kafka Extension currently only has a Release Candidate. 
+Due to this, minor releases of the extension or Axon Framework may include breaking changes to the APIs.
+
+{% endhint %}
+
+Apache Kafka is a very popular system for publishing and consuming events. 
+It's architecture is fundamentally different from most messaging systems, and combines speed with reliability.
```
Comment (YvonneCeelie) [inline extensions/kafka.md]: Its

File: extensions/kafka.md [hunk 2]
```diff
@@ -1,91 +1,437 @@
...
+
+Axon provides an extension dedicated to _publishing_ and _receiving_ event messages from Kafka.
+The Kafka Extensions should be regarded as an alternative approach to distributing events,
+ besides (the default) Axon Server.
 
-Kafka is a very popular system for publishing and consuming events. It's architecture is fundamentally different from most messaging systems, and combines speed with reliability.
+The implementation of the extension can be found in [this](https://github.com/AxonFramework/extension-kafka) repository,
```
Comment (YvonneCeelie) [inline extensions/kafka.md]: can be found here

File: extensions/kafka.md [hunk 3]
```diff
@@ -1,91 +1,437 @@
...
+To achieve this it will utilize a Kafka `Producer`, retrieved through Axon's `ProducerFactory`.
+The `KafkaPublisher` in turn receives the events to publish from a `KafkaEventPublisher`.
+
+Since the `KafkaEventPublisher` is an event message handler in Axon terms, we can provide it to any
+ [Event Processor](../configuring-infrastructure-components/event-processing/event-processors.md) to receive the
+ published events.
+The choice of event processor which brings differing characteristics for event publication to Kafka:
+
+* **Subscribing Event Processor** - publication of the messages to Kafka will happen in the same thread 
+\(and Unit of Work\) that published the events to the event bus. This approach ensures failure to send to Kafka enforces
```
Comment (YvonneCeelie) [inline extensions/kafka.md]: the same thread (and Unit of Work) as the one that .....

File: extensions/kafka.md [hunk 4]
```diff
@@ -1,91 +1,437 @@
...
+ failure of the initial event publication on the event bus
+* **Tracking Event Processor** - publication of the messages to Kafka is run in a different thread \(and Unit of Work\)
```
Comment (YvonneCeelie) [inline extensions/kafka.md]: publication of the messages to Kafka is `done`

File: extensions/kafka.md [hunk 5]
```diff
@@ -1,91 +1,437 @@
...
+ then that which published the events to the event bus.
```
Comment (YvonneCeelie) [inline extensions/kafka.md]: that -> the one

File: extensions/kafka.md [hunk 6]
```diff
@@ -1,91 +1,437 @@
...
+ [Event Processor](../configuring-infrastructure-components/event-processing/event-processors.md) to receive the
+ published events.
+The choice of event processor which brings differing characteristics for event publication to Kafka:
+
+* **Subscribing Event Processor** - publication of the messages to Kafka will happen in the same thread 
+\(and Unit of Work\) that published the events to the event bus. This approach ensures failure to send to Kafka enforces
+ failure of the initial event publication on the event bus
+* **Tracking Event Processor** - publication of the messages to Kafka is run in a different thread \(and Unit of Work\)
+ then that which published the events to the event bus. 
+ This approach ensure the event has been published on the event bus regardless of whether publication to Kafka works
```
Comment (YvonneCeelie) [inline extensions/kafka.md]: ensures

File: extensions/kafka.md [hunk 7]
```diff
@@ -1,91 +1,437 @@
...
+
+When settings up event publication it is also important to take into account which `ConfirmationMode` is used.
+The `ConfirumationMode` influences the process of actually producing an event message on a Kafka topic,
+ but also what kind of `Producer` the `ProducerFactory` will instantiate:
```
Comment (YvonneCeelie) [inline extensions/kafka.md]: `the` kind of

File: extensions/kafka.md [hunk 8]
```diff
@@ -1,91 +1,437 @@
...
 
-When Event Messages are published to an Event Bus \(or Event Store\), they can be forwarded to a Kafka topic using the `KafkaPublisher`. Publication of the messages to Kafka will happen in the same thread \(and Unit of Work\) that published the events to the Event Bus.
+* **Transactional** - This will require the `Producer` to start, commit and (in case of failure) rollback the
+ transaction of publishing an event message. 
+ Alongside this, it will create a pool of `Producer` instances in the `ProducerFactory` to avoid continuous creation of
+  new ones, requiring the user to provide a "transactional id prefix" to uniquely identify every `Producer` in the pool.
+* **Wait-for-Ack** - Setting "WAIT_FOR_ACK" as the `ConfirmationMode` will require the `Producer` instance to wait for
+ a default of 1 second (configurable on the `KafkaPublisher`) until the event message publication has ben acknowledged.
```
Comment (YvonneCeelie) [inline extensions/kafka.md]: ben -> been

File: extensions/kafka.md [hunk 9]
```diff
@@ -1,91 +1,437 @@
...
+ Alongside this, it will create a single, shareable `Producer` instance from within the `ProducerFactory`.  
+* **None** - This is the _default_ mode, which only ensures a single,
+ shareable `Producer` instance from within the `ProducerFactory`.
 
-The `KafkaPublisher` takes a `KafkaPublisherConfiguration` instance, which provides the different values and settings required to publish events to Kafka.
+### Configuring Event Publication to Kafka
+
+It is a several step process to configure Event publication to Kafka, which starts with the `ProducerFactory`.
+Axon provides a `DefaultProducerFactory` which uses the builder pattern to be instantiated.
```
Comment (YvonneCeelie) [inline extensions/kafka.md]: fwhich uses the builder pattern for the `ProducerFactory` to be instantiated

File: extensions/kafka.md [hunk 10]
```diff
@@ -1,91 +1,437 @@
...
+                .transactionalIdPrefix(transactionIdPrefix) // Hard requirement when in "TRANSACTIONAL" mode
+                .build();
+    }
+    // ...
+}
+```
 
-KafkaPublisher<String, byte[]> publisher = new KafkaPublisher<>(configuration); // create the publisher itself
+The second infrastructure component to introduce is the `KafkaPublisher`,
+ which has a hard requirement on the `ProduceFactory`.
```
Comment (YvonneCeelie) [inline extensions/kafka.md]: ProduceFactory -> ProducerFactory

File: extensions/kafka.md [hunk 11]
```diff
@@ -1,91 +1,437 @@
...
+Kafka ensures message ordering on a topic-partition level, not on an entire topic.
+To control events of a certain group, based on aggregate identifier for example, moving the a dedicated partition,
+ the [message converter's](#customizing-event-message-format) `SequencingPolicy` can be utilized.
 
-Messages can be consumed by Tracking Event Processors by configuring a `KafkaMessageSource`. This message source uses a `Fetcher` to retrieve the actual messages from Kafka. You can either use the `AsyncFetcher`, or provide your own.
+The topic-partition pair events have been published in also has impact on event consumption.
+This extension mitigates any ordering concerns with the 
+ [streamable](#consuming-events-with-a-streamable-message-source) solution,
+ by ensuring a `Consumer` always receives **all** events of a topic to be able to perform a complete ordering.
+This guarantee is however not given when using the
```
Comment (YvonneCeelie) [inline extensions/kafka.md]: This guarantee however is

File: extensions/kafka.md [hunk 12]
```diff
@@ -1,91 +1,437 @@
...
+public class KafkaEventConsumptionConfiguration {
+    // ...
+    public ConsumerFactory<String, byte[]> consumerFactory(Map<String, Object> consumerConfiguration) {
+        return new DefaultConsumerFactory<>(consumerConfiguration);
+    }
+    // ...
+}
 ```
+ 
+It is the `Fetcher` instance its job to retrieve the actual messages from Kafka by directing a `Consumer` instance it
```
Comment (YvonneCeelie) [inline extensions/kafka.md]: It is the `Fetcher` instance's job

Comment (lfgcampos) [inline extensions/kafka.md]: ~its~

File: extensions/kafka.md [hunk 13]
```diff
@@ -1,91 +1,437 @@
...
+The same benefit holds for
+ [resetting](../configuring-infrastructure-components/event-processing/event-processors.md#replaying-events) an event stream, 
+ which in Axon is reserved to the `TrackingEventProcessor`, but is now opened up through Kafka's own API's. 
 
-For further customization, you can implement your own `KafkaMessageConverter`, and wire it into the `KafkaPublisherConfiguration` and `AsyncFetcher`:
+Although the `SubscribableKafkaMessageSource` thus provides the niceties the tracking event processor normally provides,
+ it does come with two catches:
+ 
+ 1. Axon's approach of the `SequencingPolicy` to deduce which thread receives which events is entirely lost. 
+ It is thus dependent on which topic-partition pairs are given to a `Consumer` for the events your handlers receives
```
Comment (lfgcampos) [inline extensions/kafka.md]: Missing a `.`

File: extensions/kafka.md [hunk 14]
```diff
@@ -1,91 +1,437 @@
...
+ From a usage perspective this means event message ordering is no longer guaranteed by Axon. 
+ It is thus the users job to ensure events are published in the right topic-partition pair.
```
Comment (lfgcampos) [inline extensions/kafka.md]: `users' job` or `user's job` =)

File: extensions/kafka.md [hunk 15]
```diff
@@ -1,91 +1,437 @@
...
+                configuration -> subscribableKafkaMessageSource
+        );
+    }
+    // ...
+}
 ```
 
+The `KafkaMessageSourceConfigurer` is an Axon `ModuleConfiguration` which ties in to the start and end lifecycle of the application.
+It should receive the `SubscribableKafkaMessageSource` as a source which should start and stop.
+The `KafkaMessageSourceConfigurer` instance itself should be be registered as a module to the main `Configurer`.
```
Comment (lfgcampos) [inline extensions/kafka.md]: ~be~

File: extensions/kafka.md [hunk 16]
```diff
@@ -1,91 +1,437 @@
...
+
+**Generic Components:**
+ * A `DefaultKafkaMessageConverter` using the configured `eventSerializer` (which defaults to `XStreamSerializer`).
+ Uses a `String` for the keys and a `byte[]` for the record's values
+ 
+**Producer Components:**
+ * A `DefaultProducerFactory` using a `String` for the keys and a `byte[]` for the record's values.
+ This creates a `ProducerFactory` in confirmation mode "NONE", as is specified [here](#publishing-events-to-kafka).
+ The `axon.kafka.publisher.confirmation-mode` should be adjusted to change this mode,
+  where the "TRANSACTIONAL" mode requires `axon.kafka.producer.transaction-id-prefix` property should be provided.
```
Comment (lfgcampos) [inline extensions/kafka.md]: `requires ... property should be provided.` -> `requires ... property to be provided.`

File: extensions/kafka.md [hunk 17]
```diff
@@ -1,91 +1,437 @@
...
+ If the `axon.kafka.producer.transaction-id-prefix` is non-null and non-empty,
+  it is assumed a "TRANSACTIONAL" confirmation mode is desired 
+ * A `KafkaPublisher`
```
Comment (lfgcampos) [inline extensions/kafka.md]: This line and the following one is a little bit weird.

File: extensions/kafka.md [hunk 18]
```diff
@@ -1,91 +1,437 @@
...
+**Consumer Components:**
+ * A `DefaultConsumerFactory` using a `String` for the keys and a `byte[]` for the record's values
+ * An `AsyncFetcher`. To adjust the `Fetcher`'s poll timeout, the `axon.kafka.fetcher.poll-timeout` can be set.
+ * A `StreamableKafkaMessageSource` which can be used for `TrackingEventProcessor` instances
 
-To enable a `KafkaMessageSource`, either provide a bean of type `ConsumerFactory`, or provide the `axon.kafka.consumer.group-id` setting in `application.properties`. Also make sure all necessary Kafka Client Configuration properties are available under the `axon.kafka` prefix.
+When using the Spring Boot auto configuration be mindful to provide an `application.properties` file.
+The Kafka extension configuration specifics should be placed under prefix `axon.kafka`.
```
Comment (lfgcampos) [inline extensions/kafka.md]: IMO giving one example of `application.properties` with some basics would add value.

File: extensions/kafka.md [hunk 19]
```diff
@@ -1,91 +1,437 @@
...
+* **Tracking Event Processor** - publication of the messages to Kafka is run in a different thread \(and Unit of Work\)
+ then that which published the events to the event bus. 
+ This approach ensure the event has been published on the event bus regardless of whether publication to Kafka works
+
+When settings up event publication it is also important to take into account which `ConfirmationMode` is used.
+The `ConfirumationMode` influences the process of actually producing an event message on a Kafka topic,
+ but also what kind of `Producer` the `ProducerFactory` will instantiate:
 
-When Event Messages are published to an Event Bus \(or Event Store\), they can be forwarded to a Kafka topic using the `KafkaPublisher`. Publication of the messages to Kafka will happen in the same thread \(and Unit of Work\) that published the events to the Event Bus.
+* **Transactional** - This will require the `Producer` to start, commit and (in case of failure) rollback the
```
[the remark this answers is not in the document]
Comment (smcvb) [reply extensions/kafka.md]: Fair suggestion. Adjusted the fields to uppercase with underscores.

File: extensions/kafka.md [hunk 20]
```diff
@@ -1,91 +1,437 @@
...
+    // ...
+}
 ```
 
-Note that the `DefaultProducerFactory` needs to be `shutDown` properly, to ensure all producer instances are properly closed.
+### Topic partition publication considerations
 
-## Consuming Events from a Kafka topic
+Kafka ensures message ordering on a topic-partition level, not on an entire topic.
+To control events of a certain group, based on aggregate identifier for example, moving the a dedicated partition,
```
[the remark this answers is not in the document]
Comment (smcvb) [reply extensions/kafka.md]: This sentence doesn't work at all...I'll adjust it.

File: extensions/kafka.md (hunk 18 above)
Comment (smcvb) [reply extensions/kafka.md]: Sure, I'll replace the small properties snippet for an more extensive yaml snippet

**Pull request merged**